### PR TITLE
Replace LRU cache with weight-cache

### PR DESCRIPTION
**Pull request by wngr**
based on https://github.com/Actyx/weight-cache/pull/1

closes https://github.com/Actyx/banyan/issues/20

### CODE REVIEW

File: banyan/src/forest/mod.rs [hunk 1]
```diff
@@ -5,33 +5,67 @@ use anyhow::Result;
...
+use std::{fmt::Display, num::NonZeroUsize, sync::Arc};
+use weight_cache::{Weigheable, WeightCache};
 mod read;
 mod stream;
 mod write;
 pub(crate) use read::ForestIter;
 
 pub type FutureResult<'a, T> = BoxFuture<'a, Result<T>>;
 
+impl<T: TreeTypes> Weigheable for Branch<T> {
```
Comment (rklaehn) [inline banyan/src/forest/mod.rs:19]: I think this trait should be called `Weighable`. See https://en.wiktionary.org/wiki/weighable

File: banyan/src/index.rs [hunk 2]
```diff
@@ -101,6 +101,14 @@ pub trait CompactSeq: DagCbor {
     fn is_empty(&self) -> bool {
         self.len() == 0
     }
+    /// Provide a size hint, which is used to calculate the overall size of the
+    /// node, which is used for caching purposes.
+    /// This function should be overridden, if the used item type contains any
+    /// heap allocated objects, otherwise the default implementation is a rough
+    /// estimate.
+    fn size_hint(&self) -> usize {
```
Comment (rklaehn) [inline banyan/src/index.rs:109]: I would not call this size_hint, since that method for iterators gives you the number of items, not the in mem size.

Comment (rklaehn) [reply banyan/src/index.rs:109]: `estimated_size`? Not sure what a good name is. Maybe size_hint is fine after all.

Comment (wngr) [reply banyan/src/index.rs:109]: I agree that this is a potential source of confusion, so I favor `estimated_size`

**Review by rklaehn — APPROVED** [review]

File: banyan/src/forest/mod.rs [hunk 3]
```diff
@@ -5,33 +5,67 @@ use anyhow::Result;
...
 
 impl<T: TreeTypes> BranchCache<T> {
+    /// Passing a capacity of 0 disables the cache.
     pub fn new(capacity: usize) -> Self {
-        Self(Arc::new(Mutex::new(LruCache::new(capacity))))
+        let cache = if capacity == 0 {
+            None
+        } else {
+            Some(Arc::new(Mutex::new(WeightCache::new(
+                NonZeroUsize::new(capacity).expect("Cache capacity must be "),
```
Comment (rkuhn) [inline banyan/src/forest/mod.rs:55]: you probably tried this and it looks reasonable, but reading the message alone in source code feels funny :-), quite existential …

**Pull request merged**